stablehlo/reference/Ops.cpp

```diff
@@ -372,6 +403,56 @@ Tensor evalSliceOp(const Tensor &operand, Index startIndices, Sizes strides,
   return result;
 }
 
+SmallVector<Tensor> evalSortOp(ArrayRef<Tensor> inputs, Axis dimension,
+                               bool isStable, Region &comparator,
+                               Scope &scope) {
+  SmallVector<Tensor> runtimeResults;
+  for (auto input : inputs) runtimeResults.push_back(Tensor(input.getType()));
+  if (dimension < 0) dimension += inputs[0].getRank();
+  SmallVector<int64_t> indices(inputs.front().getShape()[dimension]);
+  auto inputShape = inputs.front().getShape();
+  inputShape[dimension] = 1;
+  auto inputSizes = Sizes(inputShape);
+  for (auto resultIt = IndexSpaceIterator(inputSizes, Index(inputShape.size()));
+       resultIt != IndexSpaceIterator(inputSizes, std::nullopt); ++resultIt) {
+    std::iota(indices.begin(), indices.end(), 0);
+    auto compare = [&](int64_t lhs, int64_t rhs) {
+      SmallVector<Tensor> args;
+      auto lhsIndex = *resultIt;
+      auto rhsIndex = *resultIt;
+      lhsIndex[dimension] = lhs;
+      rhsIndex[dimension] = rhs;
+      for (auto input : inputs) {
+        auto type =
+            comparator.front().getArgumentTypes()[0].dyn_cast<TensorType>();
+        auto lhs = Tensor(type);
+        auto rhs = Tensor(type);
+        lhs.set({}, input.get(lhsIndex));
+        rhs.set({}, input.get(rhsIndex));
+        args.push_back(lhs);
+        args.push_back(rhs);
+      }
+      auto cmpResult = eval(comparator, args, &scope);
+      return cmpResult[0].get({}).getBooleanValue();
+    };
+
+    if (isStable)
+      std::stable_sort(indices.begin(), indices.end(), compare);
+    else
+      std::sort(indices.begin(), indices.end(), compare);
+    auto resultIdx = *resultIt;
+    for (auto [dst, src] : llvm::enumerate(indices)) {
+      for (auto [input, result] : llvm::zip(inputs, runtimeResults)) {
+        resultIdx[dimension] = src;
+        Element resultElement = input.get(resultIdx);
+        resultIdx[dimension] = static_cast<int64_t>(dst);
+        result.set(resultIdx, resultElement);
+      }
+    }
+  }
+  return runtimeResults;
+}
+
```

Comment:
Here is your code with the suggestion mentioned above included.

```
SmallVector<Tensor> evalSortOp(ArrayRef<Tensor> inputs, Axis dimension,
                               bool isStable, Region &comparator,
                               Scope &scope) {
  SmallVector<Tensor> runtimeResults;
  for (auto input : inputs) runtimeResults.push_back(Tensor(input.getType()));

  if (dimension < 0) dimension += inputs[0].getRank();

  Index indices(inputs.front().getShape()[dimension]);

  Sizes inputShape(inputs.front().getShape());
  inputShape[dimension] = 1;
  for (auto offsetIt = IndexSpaceIterator(inputShape, Index(inputShape.size()));
       offsetIt != IndexSpaceIterator(inputShape, std::nullopt); ++offsetIt) {
    std::iota(indices.begin(), indices.end(), 0);

    auto compare = [&](int64_t lhs, int64_t rhs) {
      SmallVector<Tensor> args;
      auto lhsIndex = *offsetIt;
      auto rhsIndex = *offsetIt;
      lhsIndex[dimension] = lhs;
      rhsIndex[dimension] = rhs;
      for (const auto &indexedInput : llvm::enumerate(inputs)) {
        auto type =
            comparator.front().getArgumentTypes()[2*indexedInput.index()].dyn_cast<TensorType>();
        auto lhs = Tensor(type);
        auto rhs = Tensor(type);
        lhs.set({}, indexedInput.value().get(lhsIndex));
        rhs.set({}, indexedInput.value().get(rhsIndex));
        args.push_back(lhs);
        args.push_back(rhs);
      }
      auto cmpResult = eval(comparator, args, &scope);
      return cmpResult[0].get({}).getBooleanValue();
    };

    if (isStable)
      std::stable_sort(indices.begin(), indices.end(), compare);
    else
      std::sort(indices.begin(), indices.end(), compare);

    auto resultIdx = *offsetIt;
    auto operandIdx = *offsetIt;
    for (auto [dst, src] : llvm::enumerate(indices)) {
      for (auto [input, result] : llvm::zip(inputs, runtimeResults)) {
        operandIdx[dimension] = src;
        resultIdx[dimension] = static_cast<int64_t>(dst);
        result.set(resultIdx, input.get(operandIdx));
      }
    }
  }
  return runtimeResults;
}
```

Reply:
Thanks for the summary!